### PR TITLE
Watch tests

### PR DESCRIPTION
**Pull request by frawa**
Leverage more test-explorer UI features.

Support cancelling triggered tests.
Support autorun.
Watch file changes and notify test-explorer UI accordingly, see https://github.com/hbenl/vscode-test-adapter-api.

### CODE REVIEW

File: client/src/test-runner/adapter.ts [hunk 1]
```diff
@@ -115,26 +120,47 @@ export class ElmTestAdapter implements TestAdapter {
 
     const loadedEvent = await this.runner.runSomeTests(files);
     if (loadedEvent.suite) {
```
Comment (Razzeee) [inline client/src/test-runner/adapter.ts:122]: This if could be omitted, as the fire method handles it.

File: client/src/test-runner/adapter.ts [hunk 2]
```diff
@@ -115,26 +120,47 @@ export class ElmTestAdapter implements TestAdapter {
...
+      void this.fire(loadedEvent);
+    }
+    this.testStatesEmitter.fire(<TestRunFinishedEvent>{ type: "finished" });
+  }
+
+  private async fire(loadedEvent: TestLoadFinishedEvent): Promise<void> {
+    if (!loadedEvent.errorMessage && loadedEvent.suite) {
```
Comment (Razzeee) [inline client/src/test-runner/adapter.ts:129]: wondering, if we want to log something, in the else case

File: client/src/test-runner/adapter.ts [hunk 3]
```diff
@@ -115,26 +120,47 @@ export class ElmTestAdapter implements TestAdapter {
...
-  /*	implement this method if your TestAdapter supports debugging tests
-		async debug(tests: string[]): Promise<void> {
-			// start a test run in a child process and attach the debugger to it...
-		}
-	*/
+  private watch() {
+    this.watcher?.dispose();
+    this.watcher = undefined;
+
+    this.watcher = vscode.workspace.onDidSaveTextDocument((e) => {
```
Comment (Razzeee) [inline client/src/test-runner/adapter.ts:143]: Wondering, if we should/can filter for elm files (and maybe elm.json)

**Comment by Razzeee** [comment]
Looks good to me code wise, will try to test it for a bit.

**Comment by Razzeee** [comment]
The disable auto run setting doesn't seem to have an effect?

**Comment by frawa** [comment]
> The disable auto run setting doesn't seem to have an effect?

That depends.

When changing production code, test results will be _retired_, and test explorer run them, only if auto run is enabled.

When test code is changed, tests need to be _loaded_ into test explorer again. To detect new or removed tests.
Unfortunately, there is not way to detect available tests yet without actually running them.
That's why changing test code will always run tests, whether autorun is enabled, or not.

(I am exploring ways to load tests leveraging the Elm Language Server.)

**Comment by Razzeee** [comment]
Thank you for clearing this up.

**Pull request merged**